traits/ctraits.c

```diff
@@ -4070,10 +4071,13 @@ validate_trait_complex(
                 return result;
 
             case 22: /* Callable check: */
-                if (value == Py_None || PyCallable_Check(value)) {
-                    goto done;
+                {
+                    int allow_none = PyObject_IsTrue(PyTuple_GET_ITEM(type_info, 1));
+                    if ((allow_none && value == Py_None) || PyCallable_Check(value)) {
```

Comment:
Similar comments to those for `validate_trait_callable` apply here: we need to do an error check on the return value of `PyObject_IsTrue`, and it probably makes sense to defer the `PyObject_IsTrue` call until after checking that `value == Py_None`.